### PR TITLE
Walk / for loop over draws

### PR DESCRIPTION
**Pull request by mjskay**
This is a PR for `for_each_draw()` as per the discussion at https://github.com/stan-dev/posterior/issues/224#issuecomment-1030336782

The changes are:

1. Allows `resample_draws()` to work when `weights` is missing
2. Implement `subset_draws()`, `thin_draws()`, and `resample_draws()` for `rvar`s (#225).
3. Implements `for_each_draw(x, expr)` which acts like a for loop over draws, executing `expr` on each draw in `x`. `x` is converted to a `draws_rvars` first, so variables in `x` that represent arrays can be treated as arrays in `expr`.

Feedback here would be much appreciated. Particularly:

- ~~Does this interface solve the original problem that motivated #224?~~
  - Seems yes
- ~~Is `walk_draws()` a good name? The other name I am considering is `for_each_draw()`, which might be more widely understood ("walk" is perhaps an esoteric term).~~ 
  - Going with `for_each_draw()`

I've reproduced a version of @andrewgelman's use case here:

``` r
library(rstanarm)
library(posterior)  # use remotes::install_github("stan-dev/posterior@walk")

set.seed(4112012)

m = stan_glm(am ~ mpg, data = mtcars, family = binomial)
d = as_draws(as.matrix(m))

plot(am ~ mpg, data = mtcars)

d |>
  resample_draws(ndraws = 100) |>
  walk_draws({
    curve(invlogit(`(Intercept)` + mpg*x), col = "#FF000020", add = TRUE)
  })
```

<sup>Created on 2022-04-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Remaining TODOs:

- [x] add example in vignette
- [x] add example in function docs

### CODE REVIEW

**Comment by avehtari** [comment]
I like much much more `for_each_draw()`

**Comment by avehtari** [comment]
And this looks great, and I think it solves the problem

**Comment by mjskay** [comment]
@paul-buerkner I think this one is ready now too. Thanks!

**Comment by avehtari** [comment]
I just noticed there is `walk` in `{purr}` https://purrr.tidyverse.org/reference/index.html. I assume you know this, but why I bring this up is that  that it states
> [walk()](https://purrr.tidyverse.org/reference/map.html) is a variant for functions called primarily for their side-effects; it returns .x invisibly.

Aren't the current examples of `for_each_draw()` also called for their side effects? Maybe add to the doc of `for_each_draw()` comment about the side-effects and returns?

**Comment by mjskay** [comment]
> Aren't the current examples of for_each_draw() also called for their side effects? Maybe add to the doc of for_each_draw() comment about the side-effects and returns?

Good point! Done.

File: R/for_each_draw.R [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+#' @examples
+#' eight_schools <- as_draws_rvars(example_draws())
+#'
+#' plot(1, type = "n",
+#'   xlim = c(1, length(eight_schools$theta)),
+#'   ylim = range(range(eight_schools$theta)),
+#'   xlab = "school", ylab = "theta"
+#' )
+#'
+#' # use for_each_draw() to make a parallel coordinates plot of all draws
```
[the remark this answers is not in the document]
Comment (mjskay) [reply R/for_each_draw.R:36]: Yeah good point. I am struggling a bit to come up with a simple example that is also useful/a good idea... e.g. something like:

```r
eight_schools <- as_draws_rvars(example_draws())

for_each_draw(eight_schools, {
  print(mu)
})
```

Is simple, but also kind of pointless (and really not a good idea if people generalize that for printing rvars... in most cases it is more sensible to just print draws_of(x) if you want to see values of draws). But I could put it in there as an example with a comment say "probably don't do this".

Happy to take suggestions for a better / more useful simple example.

[the remark this answers is not in the document]
Comment (paul-buerkner) [reply R/for_each_draw.R:36]: I think this simple example is actually quite good to show whats happening and I agree adding a "probably don't do this" would be advidable to add. Thank you!

[the remark this answers is not in the document]
Comment (mjskay) [reply R/for_each_draw.R]: Cool sounds good, added!

**Comment by mjskay** [comment]
Thanks for the review --- I added the simpler example. Think it's good to go now.

**Comment by paul-buerkner** [comment]
Perfect. Thank you so much!

**Pull request merged**